minimailgun/store.py

```diff
@@ -0,0 +1,118 @@
+
+import logging
+import pymongo
+import uuid
+from collections import Mapping
+from datetime import datetime
+from pymongo.errors import DuplicateKeyError
+from minimailgun.config import config
+
+
+log = logging.getLogger(__name__)
+
+
+class MongoStore(object):
+
+    def __init__(self, hosts=None, database=None, username=None, password=None):
+        hosts = hosts or [('127.0.0.1', None)]
+        if isinstance(hosts, Mapping):
+            hosts = hosts.items()
+        hosts = ','.join([
+            '{address}:{port}'.format(address=address, port=port or '').rstrip(':')
+            for address, port in hosts
+        ])
+
+        if bool(username) != bool(password):
+            raise ValueError('username and password are mutually required.')
+        auth = '{username}:{password}@'.format(**locals()) if username else ''
+        if not database:
+            raise ValueError('database is required.')
+
+        self.url = 'mongodb://{auth}{hosts}/{database}'.format(**locals())
+        self._database = database
+        self._clear()
+
+    @property
+    def client(self):
+        if not self._client:
+            self._client = pymongo.MongoClient(self.url)
+            self.setup()
+        return self._client
+
+    @property
+    def db(self):
+        if not self._db:
+            self._db = self.client[self._database]
+        return self._db
+
+    def close(self):
+        if self._client:
+            self._client.close()
+            self._clear()
+
+    def _clear(self):
+        self._client = None
+        self._db = None
+
+    def setup(self):
+        # TODO: setup indexing here
+        pass
+
+    def add_mail(self, mail, recipient_set):
+        mail['_created_at'] = datetime.utcnow()
+        # create a recipient list and adjust the key, as bson docs
+        # don't allow key with a '.' in it.
```

Comment:
Hmm, okay - didn't know about additional characters. Initially, I was trying to keep individual documents for each recipient and even trying to keep status from all 3 tries, so could have historical status in the db and not just logs. It was getting complex and in an effort to keep this simple and admittedly, also because to time (still have to do the day job) - I did it this way.